### DISCUSSION TITLE
“Default” translation parameters

### DISCUSSION THREAD

**Discussion by martinbean** [laravel/framework#40263] [Ideas]
What do people think of specifying parameter replacements outside the normal flow of picking a string?

I’m using translations for page titles and descriptions for a sports-related site. Pretty much every string has a `:club` parameter that’s replaced with the name of a sports club. A sports club can then have multiple teams, so a lot of titles/descriptions also have a `:team` parameter. It’s therefore getting pretty cumbersome to specify these parameters for every translation, i.e.

```blade
@section('title', __(':team | Teams | :club', ['team' => $team->name, 'club' => $club->name]))
```

Especially when titles/descriptions include more and more parameters.

Therefore, my thinking is it would be good to be able to specify replacements elsewhere in an application, i.e. a controller or middleware:

```php
class SomeMiddleware
{
    public function handle($request, Closure $next)
    {
        Lang::parameter(':club', $request->route('club')->name);
        Lang::parameter(':team', $request->route('team')->name);

        return $next($request);
    }
}
```
```blade
<!-- resources/views/club/team/show.blade.php -->
@section('title', trans(':team | Teams | :club'))
<!-- i.e. "First Team | Teams | Manchester United" -->
```

In the above, the `:team` and `:club` parameters will be replaced with whatever values are set using the `Lang` façade in the example middleware class.

I’ve achieved this myself by extending the `Translator` binding in the container so already have a working solution in applications, so would also be happy to contribute the code by way of a PR and document it.

**Comment by RTippin** [comment]
I would definitely find this a useful way to clean using parameters everywhere.

**Comment by mabdullahsari** [comment]
Well, I'm going through a similar pain as we speak, so definitely a thumbs up for this.

**Comment by anbcartlow** [comment]
please do share with us how and where this can be done.
@taylorotwell should also consider this in his upcoming update for laravel
Also he should consider that there are some scenarios where hard-coded would suffice so that it can be replace in vue or js
for example: in "only ___05___ ___bats___ left. order soon", both params need replacement in vuejs but we also use them in laravel.
in laravel I can while calling trans() provide params but when an array of keys with params is to be passed to a vue file from blade, its hard to provide params. we need a way for it too

**Comment by Joel-Jensen** [comment]
For anyone else who ends up here, a quickfix that worked for me until this is a part of the core:

_AppServiceProvider.php_
```php
public function register() {
    $this->app->extend('translator', function (Translator $translator) {
        return new DefaultParamTranslator($translator->getLoader(), $translator->getLocale());
    });
}
```

_DefaultParamTranslator.php_
```php
use Illuminate\Translation\Translator;

class DefaultParamTranslator extends Translator {

    public function get($key, array $replace = [], $locale = null, $fallback = true)
    {
        return parent::get($key, [
            'company_name' => 'Laravel',
            ...$replace
        ], $locale, $fallback);
    }

}
```

Open for better suggestions, quite new to container bindning
